src/util/LazyJsonParser.cpp

```diff
@@ -0,0 +1,258 @@
+      throw std::runtime_error(
+          "Ill-formed JSON: Element size exceeds 1 Megabyte.");
+    }
+    switch (input_[idx]) {
+      case '{':
+      case '[':
+        ++state.openBracketsAndBraces_;
+        break;
+      case '}':
+        --state.openBracketsAndBraces_;
+        break;
+      case ']':
+        if (state.openBracketsAndBraces_ == 0) {
+          // End of ArrayPath reached.
+          state_ = AfterArrayPath{.remainingBraces_ = arrayPath_.size(),
+                                  .arrayEndIdx_ = idx};
+          ++idx;
+          return;
+        }
+        --state.openBracketsAndBraces_;
+        break;
+      case ',':
+        if (state.openBracketsAndBraces_ == 0) {
+          materializeEnd = idx;
+          // reset arrayStartIdx_ to its sentinel 0, as it's only needed/valid
+          // for the first element
+          state.arrayStartIdx_ = 0;
+        }
+        break;
+      case '"':
+        parseLiteral(idx);
+        break;
+      default:
+        break;
+    }
+  }
+}
+
+// ____________________________________________________________________________
+void LazyJsonParser::parseAfterArrayPath(size_t& idx, size_t& materializeEnd) {
+  AD_CORRECTNESS_CHECK(std::holds_alternative<AfterArrayPath>(state_));
+  auto& state = std::get<AfterArrayPath>(state_);
+
+  for (; idx < input_.size(); ++idx) {
+    if (idx - state.arrayEndIdx_ > 1'000'000) {
+      throw std::runtime_error(
+          "Ill-formed JSON: Element size exceeds 1 Megabyte.");
+    }
+    switch (input_[idx]) {
+      case '{':
+        state.remainingBraces_ += 1;
+        break;
+      case '}':
+        state.remainingBraces_ -= 1;
+        if (state.remainingBraces_ == 0) {
+          // End reached.
+          materializeEnd = idx + 1;
+        }
+        break;
+      case '"':
+        parseLiteral(idx);
+        break;
+      default:
+        break;
+    }
+  }
+}
+
+// ____________________________________________________________________________
+std::string LazyJsonParser::constructResultFromParsedChunk(
+    size_t materializeEnd) {
+  std::string res;
+  if (materializeEnd == 0) {
+    return res;
+  }
+
+  // Don't yield before arrayPath.
+  AD_CORRECTNESS_CHECK(!std::holds_alternative<BeforeArrayPath>(state_));
+
+  if (yieldCount_ > 0) {
+    res = prefixInArray_;
+  }
+  ++yieldCount_;
+
+  absl::StrAppend(&res, input_.substr(0, materializeEnd));
+  input_ = input_.substr(std::min(materializeEnd + 1, input_.size()));
+
+  if (std::holds_alternative<InArrayPath>(state_)) {
+    absl::StrAppend(&res, suffixInArray_);
+  }
+
+  return res;
+}
+
+// ____________________________________________________________________________
+void LazyJsonParser::BeforeArrayPath::tryAddKeyToPath(std::string_view input) {
+  if (litLength_ > 0) {
+    curPath_.emplace_back(input.substr(litStart_, litLength_));
+    // The marked literal got consumed/added as key -> reset litLength_.
+    litLength_ = 0;
```

Comment:
Actually: Are the `""` part of the literal or not?
Otherwise your case `litLength_ == 0` is ambigous (no literal vs. empty literal).
I think the much cleaner interface is to store the
`litStart_ and litLenght_` in a struct and wrap this in an `std::optional<>` s.t. you really have a clear "there just was no literal" case.